lib/mlir/Pipes/MLIRContainer.cpp

```diff
@@ -0,0 +1,270 @@
+      for (const mlir::SymbolTable::SymbolUse &Use : *Uses) {
+        const auto &SymbolRef = Use.getSymbolRef();
+        revng_assert(SymbolRef.getNestedReferences().empty());
+
+        mlir::Operation *const
+          Symbol = mlir::SymbolTable::lookupSymbolIn(Module,
+                                                     SymbolRef
+                                                       .getRootReference());
+        revng_assert(Symbol);
+
+        UsedSymbols.insert(Symbol);
+      }
+    }
+  });
+
+  visit(Module, [&](mlir::SymbolOpInterface S) {
+    if (not UsedSymbols.contains(S.getOperation()))
+      S->erase();
+  });
+}
+
+const char MLIRContainer::ID = 0;
+
+void MLIRContainer::setModule(mlir::OwningOpRef<mlir::ModuleOp> &&NewModule) {
+  revng_assert(NewModule);
+
+  // Make any non-target functions external.
+  visit(*NewModule, [&](mlir::FunctionOpInterface F) {
+    if (not F.isExternal() and not getFunctionName(F))
+      makeExternal(F);
+  });
+
+  pruneSymbols(*NewModule);
+  Module = std::move(NewModule);
+}
+
+std::unique_ptr<pipeline::ContainerBase>
+MLIRContainer::cloneFiltered(const pipeline::TargetsList &Filter) const {
+  auto NewContainer = std::make_unique<MLIRContainer>(name());
+  if (not getModuleBlock(*Module).empty()) {
+    OwningModuleRef NewModule(mlir::cast<mlir::ModuleOp>((*Module)->clone()));
+
+    bool RemovedSome = false;
+    visit(*NewModule, [&](mlir::FunctionOpInterface F) {
+      if (F.isExternal())
+        return;
+
+      const auto MaybeTargetName = getFunctionName(F);
+      if (not MaybeTargetName)
+        return;
+
+      if (not Filter.contains(makeTarget(*MaybeTargetName))) {
+        makeExternal(F);
+        RemovedSome = true;
+      }
+    });
+
+    if (RemovedSome)
+      pruneSymbols(*NewModule);
+
+    NewContainer->Context->appendDialectRegistry(Context->getDialectRegistry());
+    NewContainer->Module = cloneModuleInto(*NewModule, *NewContainer->Context);
+  }
+  return NewContainer;
+}
+
+void MLIRContainer::mergeBackImpl(MLIRContainer &&Container) {
+  if (getModuleBlock(*Container.Module).empty())
+    return;
+
+  if (getModuleBlock(*Module).empty()) {
+    Module = std::move(Container.Module);
+    Context = std::move(Container.Context);
+    return;
+  }
+
+  // Register the dialects of the other container in this container.
+  Context->appendDialectRegistry(Container.Context->getDialectRegistry());
+
+  // Clone the other container's module into this container's context.
+  // This module is automatically erased at the end of scope.
+  OwningModuleRef TemporaryModule = cloneModuleInto(*Container.Module,
+                                                    *Context);
+
+  mlir::Block &Block = getModuleBlock(*Module);
+  visit(*TemporaryModule, [&](mlir::SymbolOpInterface Symbol) {
+    // Erase an existing symbol with the same name, if one exists.
+    if (auto S = mlir::SymbolTable::lookupSymbolIn(*Module, Symbol.getName()))
+      S->erase();
```

Comment:
As in the other comment, AFAIK `clift.function`s won't implement the `Symbol` interface.